smac/intensification/successive_halving.py

```diff
@@ -0,0 +1,475 @@
+                         deterministic=deterministic, run_obj_time=run_obj_time,
+                         adaptive_capping_slackfactor=adaptive_capping_slackfactor)
+
+        self.logger = logging.getLogger(
+            self.__module__ + "." + self.__class__.__name__)
+
+        # INSTANCES
+
+        self.n_seeds = n_seeds if n_seeds else 1
+        self.instance_order = instance_order
+
+        # if instances are coming from Hyperband, skip the instance preprocessing section
+        # it is already taken care by Hyperband
+        if instances is not None and isinstance(instances[0], InstSeedKey):
+            self.instances = instances
+        else:
+            instances = [] if instances is None else instances
+            # removing duplicates in the user provided instances
+            self.instances = list(OrderedDict.fromkeys(instances))
+
+            # determine instance order
+            if self.instance_order == 'shuffle_once':
+                # randomize once
+                self.rs.shuffle(self.instances)
+
+            # set seed(s) for all SH runs
+            # - currently user gives the number of seeds to consider
+            if self.deterministic:
+                seeds = [0]
+            else:
+                seeds = self.rs.randint(low=0, high=MAXINT, size=self.n_seeds)
+            # storing instances & seeds as tuples
+            self.instances = [InstSeedKey(i, s) for s in seeds for i in self.instances]
+
+        # successive halving parameters
+        self._init_sh_params(min_budget, max_budget, eta, init_chal)
+
+        # adaptive capping
+        if not self.cutoff_as_budget and self.instance_order != 'shuffle' and self.run_obj_time:
+            self.adaptive_capping = True
+        else:
+            self.adaptive_capping = False
+
+    def _init_sh_params(self, min_budget: float,
+                        max_budget: float,
+                        eta: float,
+                        init_chal: int):
+        """
+        initialize Successive Halving parameters
+
+        Parameters
+        ----------
+        min_budget : float
+            minimum budget allowed for 1 run of successive halving
+        max_budget : float
+            maximum budget allowed for 1 run of successive halving
+        eta : float
+            'halving' factor after each iteration in a successive halving run
+        init_chal : int
+            number of challengers to consider for the initial budget
+        """
+
+        if eta <= 1:
+            raise ValueError('eta must be greater than 1')
+        self.eta = eta
+
+        # BUDGETS
+
+        # - if only 1 instance was provided & quality objective, then use cutoff as budget
+        # - else, use instances as budget
+        if not self.run_obj_time and len(self.instances) <= 1:
+            # budget with cutoff
+            # cannot run successive halving with cutoff as budget if budget limits are not provided!
+            if min_budget is None or \
```

Comment:
Above you define them as float, while here you also allow them to be `None`. Please update the method signature to reflect this.